app/src/main/java/com/android/sample/ui/filter/FilterPage.kt

```diff
@@ -0,0 +1,317 @@
+      Text(
+          "${sliderPosition.start.toInt()} " + unit,
+          style = MaterialTheme.typography.bodyMedium,
+          color = MaterialTheme.colorScheme.onPrimary)
+      Text(
+          "${sliderPosition.endInclusive.toInt()} " + unit,
+          style = MaterialTheme.typography.bodyMedium,
+          color = MaterialTheme.colorScheme.onPrimary)
+    }
+
+    Spacer(modifier = Modifier.height(8.dp))
+
+    // Range slider with minimum 0 and maximum 120 minutes
+    RangeSlider(
+        modifier = modifier.fillMaxWidth(),
+        value = sliderPosition,
+        onValueChange = {
+          sliderPosition = it
+          updateRange(it.start, it.endInclusive)
+        },
+        valueRange = min..max,
+        colors =
+            SliderDefaults.colors(
+                thumbColor = MaterialTheme.colorScheme.primary,
+                activeTrackColor = MaterialTheme.colorScheme.primary.copy(alpha = 0.4f),
+                inactiveTrackColor = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.2f)))
+  }
+}
+
+/**
+ * Composable function to display checkboxes for difficulty levels.
+ *
+ * @param recipesViewModel The view model to manage recipes.
+ */
+@SuppressLint("MutableCollectionMutableState", "StateFlowValueCalledInComposition")
+@OptIn(ExperimentalLayoutApi::class)
+@Composable
+fun CheckboxDifficulty(recipesViewModel: RecipesViewModel) {
+  // List of difficulty names
+
+  val difficultyLevels = remember {
+    mutableStateMapOf(
+        Difficulty.Easy to false, Difficulty.Medium to false, Difficulty.Hard to false)
+  }
+  val orderedDifficulties = listOf(Difficulty.Easy, Difficulty.Medium, Difficulty.Hard)
+
+  if (recipesViewModel.filter.value.difficulty != Difficulty.Undefined) {
+    difficultyLevels[recipesViewModel.filter.value.difficulty] = true
+  }
+
+  Column(modifier = Modifier.padding(16.dp), horizontalAlignment = Alignment.CenterHorizontally) {
+    Text(
+        text = "Difficulty",
+        style = MaterialTheme.typography.titleMedium,
+        modifier = Modifier.align(Alignment.Start))
+
+    // Display each checkbox with the corresponding difficulty name
+    FlowRow(
+        modifier = Modifier.fillMaxWidth(),
+        horizontalArrangement = Arrangement.Center,
+        verticalArrangement = Arrangement.Center,
+        maxItemsInEachRow = 3) {
+          orderedDifficulties.forEach { difficulty ->
+            Row(
+                verticalAlignment = Alignment.CenterVertically,
+            ) {
+              Spacer(modifier = Modifier.height(4.dp))
+
+              Text(
+                  text = difficulty.toString(),
+                  style = MaterialTheme.typography.bodyMedium,
+                  color = MaterialTheme.colorScheme.onPrimary) // Display the difficulty name
+              Checkbox(
+                  modifier = Modifier.testTag("difficultyCheckbox${difficulty}"),
+                  checked = difficultyLevels[difficulty]!!,
+                  onCheckedChange = { isChecked ->
```

Comment:
The `CheckboxDifficulty` and `CheckboxCategories` methods have a lot of overlapping code, particularly in how the checkboxes and categories are displayed. It might be worth modularizing this shared logic into a single reusable component. This would help streamline updates, improve readability, and maintain consistency across both methods. Great work overall, just a small suggestion to enhance maintainability!"

Reply:
It was a bit challenging, but I managed to complete it!